### DISCUSSION TITLE
How to install Bamcmp in a modularized HPC environment

### DISCUSSION THREAD

**Discussion by Ravnell** [community/community#29953] [Other Feature Feedback, Questions, & Ideas]
Not sure if this is the correct place to ask this but I'm in need of instruction on how to install this software in our HPC.  I've setup the folder structure and modified the Makefile as I thought was needed to reflect the install location but constantly get errors trying to compile.

/usr/bin/ld: cannot find -lhts
collect2: error: ld returned 1 exit status
make: *** [bamcmp] Error 1

Here's my Makefile mods:

# bamcmp Makefile

ifndef HTSLIBDIR
   HTSLIBDIR=/apps/htslib/1.13/
endif

CC=gcc
CFLAGS=-g -Wall-fexceptions
CPP=g++
CPPFLAGS=-g -Wall -fexceptions
LDFLAG=-g
LDLIBS=hts

SRCDIR=/apps/bamcmp/2.2/src/src/
INCDIR=/apps/bamcmp/2.2/src/include/
BUILDDIR=/apps/bamcmp/2.2/src/build/
SRCS=SamReader.cpp BamRecVector.cpp HTSFileWrapper.cpp util.cpp bamcmp.cpp
OBJS=$(BUILDDIR)SamReader.o $(BUILDDIR)BamRecVector.o \
        $(BUILDDIR)HTSFileWrapper.o $(BUILDDIR)util.o $(BUILDDIR)bamcmp.o

bamcmp: ${OBJS} $(BUILDDIR)
        $(CPP) $(LDFLAGS) -o $(BUILDDIR)/bamcmp $(OBJS) -L $(HTSLIBDIR) -l $(LDLIBS) -Wl,-rpath,/apps/bamcmp/2.2/

$(BUILDDIR)%.o: $(SRCDIR)%.cpp $(BUILDDIR)
        $(CPP) $(CPPFLAGS) -I $(INCDIR) -I $(HTSLIBDIR)/include -o $@ -c $<

$(BUILDDIR):
        mkdir $(BUILDDIR)

clean:
        rm -f $(OBJS) $(BUILDIR)/bamcmp

I definitely don't know what I'm doing and wondered if anyone could help?

Thanks.

**Comment by airtower-luna** [comment]
It looks like you're looking for help with a software hosted on GitHub, maybe https://github.com/CRUKMI-ComputationalBiology/bamcmp? In that case you should open an issue *in that repository* so the developers see it.

However from my general C knowledge the error looks like you're missing the required `htslib` (or don't have the path to it set correctly), so you should check that first. :wink:

**Reply by Ravnell** [reply]
Thanks for the reply.  Since we use a modularize app environment, I have to load the htslib library beforehand.  

[username@servername src]$ module list
Currently Loaded Modulefiles:
 1) gcc/8.2.0   2) slurm/18.08.9   3) htslib/1.13

If I do not load it, I get:

[username@servername src]$ make
g++ -g -Wall -fexceptions -I /apps/bamcmp/2.2/src/include/ -I /apps/htslib/1.13//include -o /apps/bamcmp/2.2/src/build/SamReader.o -c /apps/bamcmp/2.2/src/src/SamReader.cpp
In file included from /apps/bamcmp/2.2/src/src/SamReader.cpp:21:
/apps/bamcmp/2.2/src/include/SamReader.h:5:10: fatal error: htslib/sam.h: No such file or directory
 #include <htslib/sam.h>
          ^~~~~~~~~~~~~~
compilation terminated.
make: *** [/apps/bamcmp/2.2/src/build/SamReader.o] Error 1

So it should be available to the complier in my thinking.  Again, I'm fairly new at installing apps in this type of environment.

Again, thank you very much for your response.  I was trying to place my question there but I've only used github for downloading.  My first time asking for assistance and thus apologize for posting in the wrong area.

**Comment by airtower-luna** [comment]
That is a compiler error, header can't be found:

> g++ -g -Wall -fexceptions -I /apps/bamcmp/2.2/src/include/ -I /apps/htslib/1.13//include -o /apps/bamcmp/2.2/src/build/SamReader.o -c /apps/bamcmp/2.2/src/src/SamReader.cpp
> In file included from /apps/bamcmp/2.2/src/src/SamReader.cpp:21:
/apps/bamcmp/2.2/src/include/SamReader.h:5:10: fatal error: htslib/sam.h: No such file or directory
> #include <htslib/sam.h>
> ^~~~~~~~~~~~~~
> compilation terminated.

What you had above was a *linker* error, it can't find `libhts`:

> /usr/bin/ld: cannot find -lhts

You need to provide both headers and the dynamic library, and need to (separately) set paths for both (`-I` for header search path, `-L` for library search path). You also need to consider how the dynamically linked library will be found at runtime (there are multiple ways to do this).

> My first time asking for assistance and thus apologize for posting in the wrong area.

It's less "wrong area" and more "wrong town". The GitHub Community is mostly for general discussions about Git and GitHub, of course a lot of people here know programming in some way (and it's perfectly fine to discuss here!). The repository is the specific place *for that software*. What you did is kind of like having trouble with your rare model of car and going to a DIY meetup and asking random people, instead of going to a specialized workshop. You'll probably find someone who has some idea about cars, but it's unlikely someone knows your unusual model. Like I know about compiler and linker errors, but I have no idea about Bamcmp or your modular environment. :wink:

**Comment by Ravnell** [comment]
That was very informative.  This gives me more information to work from.

I loved the analogy and once again thank you for your time and the helpful response.

I hope you have a great rest of the week!
